homeassistant/components/mysensors/config_flow.py

```diff
@@ -0,0 +1,176 @@
+            is_mqtt = user_input[CONF_DEVICE] == MQTT_COMPONENT
+            is_serial = False
+            try:
+                if user_input[CONF_DEVICE] != MQTT_COMPONENT:
+                    try:
+                        await self.hass.async_add_executor_job(
+                            is_serial_port, user_input[CONF_DEVICE]
+                        )
+                        is_serial = True
+                    except vol.Invalid:
+                        await self.hass.async_add_executor_job(
+                            is_socket_address, user_input[CONF_DEVICE]
+                        )
+                        is_serial = False
+            except vol.Invalid:
+                errors[CONF_DEVICE] = "invalid_device"
+
+            try:
+                if is_mqtt and CONF_TOPIC_IN_PREFIX in user_input:
+                    valid_subscribe_topic(user_input[CONF_TOPIC_IN_PREFIX])
+            except vol.Invalid:
+                errors[CONF_TOPIC_IN_PREFIX] = "invalid_subscribe_topic"
+            try:
+                if is_mqtt and CONF_TOPIC_OUT_PREFIX in user_input:
+                    valid_publish_topic(user_input[CONF_TOPIC_OUT_PREFIX])
+            except vol.Invalid:
+                errors[CONF_TOPIC_OUT_PREFIX] = "invalid_publish_topic"
+            if not is_mqtt and (
+                CONF_TCP_PORT in user_input
+                and (user_input[CONF_TCP_PORT] < 1 or user_input[CONF_TCP_PORT] > 65535)
+            ):
+                errors[CONF_TCP_PORT] = "invalid_port"
+            try:
+                if CONF_PERSISTENCE_FILE in user_input:
+                    is_persistence_file(user_input[CONF_PERSISTENCE_FILE])
+            except vol.Invalid:
+                errors[CONF_TCP_PORT] = "invalid_persistence_file"
+
+            uniquestr = user_input[CONF_DEVICE]
+            if is_mqtt:
+                uniquestr += (
+                    user_input[CONF_TOPIC_IN_PREFIX] + user_input[CONF_TOPIC_OUT_PREFIX]
+                )
+            elif not is_serial:
+                uniquestr += str(user_input[CONF_TCP_PORT])
+            gateway_id = hashlib.sha256(uniquestr.encode()).hexdigest()[:8]
+            await self.async_set_unique_id(gateway_id)
+            self._abort_if_unique_id_configured()
+
+            # if no errors so far, try to connect
+            if not errors and not await try_connect(
+                self.hass, user_input, uniqueid=gateway_id
+            ):
+                errors["base"] = "cannot_connect"
+
+            if not errors:
+                _LOGGER.info("config_flow completed for %s", gateway_id)
+                return self.async_create_entry(
+                    title=f"{user_input[CONF_DEVICE]}", data=user_input
+                )
+
+        schema = OrderedDict()
+        schema[vol.Optional(CONF_OPTIMISTIC, default=False)] = bool
+        schema[vol.Optional(CONF_PERSISTENCE, default=True)] = bool
+        schema[vol.Optional(CONF_RETAIN, default=True)] = bool
```

Comment:
This options should only be shown for MQTT gateways. Do a two step form flow where first selecting gateway type and then the gateway options required to connect.